### PR TITLE
[#1902]: Fix reflection

### LINKED ISSUE #1902: Cache doesn't find `getMetadata` method although it is present
The `Cache` implementation doesn't find the `getMetadata` method although it is present:

~~~
2019-12-10T14:32:21.596Z ERROR [ReflectorRunnable] Failure in list-watch: java.lang.ReflectiveOperationException: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
java.lang.RuntimeException: java.lang.ReflectiveOperationException: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
	at io.fabric8.kubernetes.client.informers.cache.Cache.metaNamespaceKeyFunc(Cache.java:365)
	at io.fabric8.kubernetes.client.informers.cache.DeltaFIFO.keyOf(DeltaFIFO.java:381)
	at io.fabric8.kubernetes.client.informers.cache.DeltaFIFO.replace(DeltaFIFO.java:165)
	at io.fabric8.kubernetes.client.informers.cache.ReflectorRunnable.syncWith(ReflectorRunnable.java:145)
	at io.fabric8.kubernetes.client.informers.cache.ReflectorRunnable.run(ReflectorRunnable.java:69)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ReflectiveOperationException: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
	at io.fabric8.kubernetes.client.utils.ReflectUtils.objectMetadata(ReflectUtils.java:31)
	at io.fabric8.kubernetes.client.informers.cache.Cache.metaNamespaceKeyFunc(Cache.java:355)
	... 10 more
Caused by: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2475)
	at io.fabric8.kubernetes.client.utils.ReflectUtils.objectMetadata(ReflectUtils.java:28)
	... 11 more
~~~

This is due to the fact that `ReflectUtils` uses `getDeclaredMethod`, which will only find methods which are directly declared on the class:

https://github.com/fabric8io/kubernetes-client/blob/e2a8c9f7b9329099b1ef88e3f1d27f7a4056ab71/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/ReflectUtils.java#L28-L29

Instead of doing reflection, the code should check for `instanceof HasMetadata` and maybe then use `Class#getMethod`.

### PR DESCRIPTION
**Pull request by ctron**
Fixes #1902

### CODE REVIEW

**Comment by rohanKanojia** [comment]
ok to test

**Comment by rohanKanojia** [comment]
@ctron : Thanks for the PR :+1: . Could you please add a line into `CHANGELOG.md` and correct license using `mvn -N license:format`?

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/ReflectUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      return Baz.class.getSimpleName();
+    }
+  }
+
+  @Test
+  public void testDirectReflection () throws ReflectiveOperationException {
+    ObjectMeta meta = new ObjectMeta();
+    Foo foo = new Foo();
+    foo.setMetadata(meta);
+    assertTrue(ReflectUtils.objectMetadata(foo) == meta);
```
Comment (manusa) [inline kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/ReflectUtilsTest.java]: Can be simplified to `assertSame`

**Comment by ctron** [comment]
Thanks for the review. I ran the license plugin (although I think the copyright year is wrong now) and changed to use `assertSame`.

**Comment by rohanKanojia** [comment]
retest this please

**Comment by rohanKanojia** [comment]
retest this please

**Comment by rohanKanojia** [comment]
retest this please

**Comment by rohanKanojia** [comment]
retest this please

**Comment by rohanKanojia** [comment]
retest this please

**Comment by rohanKanojia** [comment]
[merge]

**Pull request merged**